Add --use-dummy parameter to support VMA's Dummy send feature.

liranoz12 opened the pull request:
WARNING - Using this parameter without VMA will cause all packets to be
sent.

Signed-off-by: Liran Oz <lirano@mellanox.com>

Commit:
Add --use-dummy parameter to support VMA's Dummy send feature.

WARNING - Using this parameter without VMA will cause all packets to be
sent.

Signed-off-by: Liran Oz <lirano@mellanox.com>

Commit:
Some changes:

1. Rename --use-dummy to --dummy-send.
2. Enable dummy-send only at client side.

Signed-off-by: Liran Oz <lirano@mellanox.com>

Commit:
Fix AlexR review comments:

1. Add DUMMY_SEND_FLAG flag
2. SwitchOnCycleDuration class - use default values in execute()
3. SwitchOnDummySend class - use only one while loop in execute()
4. Set dummy-send rate per second (default 10000). usage: --dummy-send <rate>

Signed-off-by: Liran Oz <lirano@mellanox.com>

rosenbaumalex commented on src/Switches.h:
```diff
@@ -94,7 +95,35 @@ class SwitchOnCycleDuration {
 		g_cycleStartTime = nextCycleStartTime;
 	}
+
+	inline void execute(int ifd, int buffer_size) {
```
1. Better to use the original function with default arguments so you can call with and without input values.
2. Also add something line "NOT_IN_USE(x)" marco. Is there something like that in sockperf today?

liranoz12 replied:
1. Yes i agree.
2. No, Sockperf does not have something like "NOT_IN_USE(x)" and i don't sure it is necessary. There are many areas like this in the code :
```
class SwitchOff {
public:
	inline void execute() {}
	inline void execute(int) {}
//	inline void execute(unsigned long long) {}
	inline void execute(uint64_t) {}
	inline void execute(int, uint64_t) {}
	inline void execute(TicksTime & _ticks) {}
	inline void execute(struct sockaddr_in *clt_addr, uint64_t seq_num, bool is_warmup) {}
	inline void execute(Message *pMsgRequest, Message * pMsgReply) {}
	inline void execute(Message *pMsgRequest) {}
};
```

rosenbaumalex commented on src/Switches.h:
```diff
@@ -94,7 +95,35 @@ class SwitchOnCycleDuration {
+				break;
+			}
+
+			sendto(ifd, buffer, sizeof(buffer), MSG_SYN, (struct sockaddr*)& (g_fds_array[ifd]->server_addr), sizeof(struct sockaddr));
```
Currently this dummy send is happening at the max rate allowed which is not  what we expect customers to do in real life. We don't want to send in such a high rate.

Better to have the '--use-dummy' configuration to get a rate, so we can run it with 1000 mps as warmup.
I would recommend having a default rate of 1000mps as --dummy-send-rate, and optionally have user set a different value, like 10kmps

rosenbaumalex replied:
better to create a sockperf level define, something like:
#define DUMMY_SEND_FLAG    MSG_SYN // Used by offload libraries to do egress path warm-up of caches. It is not in use by kernel. WARNING: it will actually end this packet on the wire.

rosenbaumalex replied:
After more thought, a bigger design change...
I think the "send dummy" should be separate from the sleep logic. lets talk about it offline.

rosenbaumalex reviewed (changes requested):
After more thought, I want to discuss with you a bigger design change... I think the "send dummy" should be separate from the sleep logic. let's talk about it offline.

liranoz12 commented:
@rosenbaumalex  Please review

rosenbaumalex commented on src/Client.h:
```diff
@@ -313,7 +313,7 @@ class Client : public ClientBase{
 
 		//idle
-		m_switchCycleDuration.execute();
+		m_switchCycleDuration.execute(ifd, m_pMsgRequest->getLength());
```
don't pass lenght, pass the m_pMsg to execute()

rosenbaumalex commented on src/Defs.h:
```diff
@@ -147,6 +147,8 @@ extern const int MAX_FDS_NUM;
 #define DEFAULT_SELECT_TIMEOUT_MSEC	10
 #define DEFAULT_DEBUG_LEVEL			0
 
+#define DUMMY_SEND_FLAG MSG_SYN // Used by offload libraries to do egress path warm-up of caches. It is not in use by kernel. WARNING: it will actually end this packet on the wire.
```
1.  break comment to NEWLINEs
2. read more in URL.. to github/libvma/... vma_extra.h

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -312,6 +312,10 @@ static const AOPT_DESC  client_opt_desc[] =
 	},
+	{
+		OPT_DUMMY_SEND, AOPT_OPTARG,	aopt_set_literal( 0 ),	aopt_set_string( "dummy-send" ),
+		"Use dummy-send instead of busy wait. optional: set dummy-send rate per second (default 10000). usage: --dummy-send <rate>"
```
fix:
1. 10,000 (use comma)
2. option: "[<rate>]"
3. add "MAX" as option
4. explain that rate value "must be lower than --mps"

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -2093,6 +2097,26 @@ static int parse_client_opt( const AOPT_OBJECT *client_obj )
 			s_user_params.increase_output_precision = true;
 		}
+		if ( !rc && aopt_check(client_obj, OPT_DUMMY_SEND) ) {
+			s_user_params.dummy_send = true;
```
check if can get rid of bool, and use !=0 fro ON and 0 for OFF

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -2290,6 +2314,8 @@ void set_defaults()
 
 	s_user_params.withsock_accl = false;
+	s_user_params.dummy_send = false;
+	s_user_params.dummy_mps = DUMMY_SEND_MPS_DEFAULT;
```
move set of default to only if command line enabled

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -3205,6 +3231,16 @@ int bringup(const int *p_daemonize)
+		}
+
+		if ( !rc && s_user_params.dummy_send && s_user_params.mps > s_user_params.dummy_mps) {
+			log_err("Dummy send is not supported when message-rate is greater then the dummy-message-rate");
```
error msg: "Dummy send is allowed only if dumm-send-rate is lower than regular msg rate"

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -3205,6 +3231,16 @@ int bringup(const int *p_daemonize)
 			rc = SOCKPERF_ERR_BAD_ARGUMENT;
 		}
+
+		if ( !rc && s_user_params.dummy_send && s_user_params.mps == UINT32_MAX) {
```
will be removed once bool is removed

rosenbaumalex commented on src/SockPerf.cpp:
```diff
@@ -3205,6 +3231,16 @@ int bringup(const int *p_daemonize)
+			rc = SOCKPERF_ERR_BAD_ARGUMENT;
+		}
+
+		if ( !rc && s_user_params.dummy_send && s_user_params.mps > s_user_params.dummy_mps) {
```
use >= instead of > so dummy rate must be slightly lower the packet rate ALWAYS

rosenbaumalex commented on src/Switches.h:
```diff
@@ -47,6 +47,7 @@ class SwitchOff {
 	inline void execute(struct sockaddr_in *clt_addr, uint64_t seq_num, bool is_warmup) {}
 	inline void execute(Message *pMsgRequest, Message * pMsgReply) {}
 	inline void execute(Message *pMsgRequest) {}
+	inline void execute(int ifd, int buffer_size) {}
```
instead of buf_size use pMsgRequest

rosenbaumalex commented on src/Switches.h:
```diff
@@ -95,6 +96,33 @@ class SwitchOnCycleDuration {
+			}
+
+			if (TicksTime::now() > nextDummySendTime) {
+				sendto(ifd, buffer, sizeof(buffer), DUMMY_SEND_FLAG, (struct sockaddr*)& (g_fds_array[ifd]->server_addr), sizeof(struct sockaddr));
```
we should consider using a flag which is not supported by OS so this frame gets dropped and does not get queued by OS on TCP and UDP flows.

rosenbaumalex reviewed (commented):
1. some small fixes
2. consider using un=-supported flag so send will fail

Commit:
Fix AlexR review comments:

1. Don't pass lenght, pass the m_pMsg to execute()
2. Update DUMMY_SEND_FLAG comment
3. Update dummy-send section in sockperf help.
4. Remove user_params.dummy_send boolean.
5. Dummy rate must be slightly higher the packet rate (also update the error message)

Signed-off-by: Liran Oz <lirano@mellanox.com>

liranoz12 commented:
@rosenbaumalex Please review

avnerbh commented on src/Switches.h:
```diff
@@ -83,7 +84,7 @@ class SwitchOnMsgSize {
 public:
 	//busy wait between two cycles starting point and take starting point of next cycle
-	inline void execute() {
+	inline void execute(Message *pMsgRequest, int ifd) {
```
it is better to replace:
  

> execute(Message *pMsgRequest, int ifd)

with:
> execute(Message *, int)

(and if you want, feel free to also fix  above)

avnerbh commented on src/Switches.h:
```diff
@@ -95,6 +96,33 @@ class SwitchOnCycleDuration {
+	// dummy send between two cycles starting point and take starting point of next cycle
+
+	inline void execute(Message *pMsgRequest, int ifd) {
+		uint8_t buffer[pMsgRequest->getLength()];
```
*this line is not C code*, but a gcc extension - hence, you must make sure it compiles on all supported platforms, like Windows and BSD 
if there is no additional example in SockPerf for this trick, I recommend that you will use static array of size MAX_MSG_LEN

avnerbh commented on src/Defs.h:
```diff
@@ -147,6 +147,13 @@ extern const int MAX_FDS_NUM;
+It is not in use by kernel. WARNING: it will actually end this packet on the wire.
+More info at vma_extra.h
+*/
+#define DUMMY_SEND_FLAG MSG_SYN // equals to 0x400
```
it can be safer to add a comment - this value is shared with vma (and solarflare); hence, don't change here without changing accordingly in VMA

avnerbh commented on src/Switches.h:
```diff
@@ -95,6 +96,33 @@ class SwitchOnCycleDuration {
+		TicksTime nextCycleStartTime = g_cycleStartTime + g_pApp->m_const_params.cycleDuration;
+		TicksTime nextDummySendTime = g_cycleStartTime + g_pApp->m_const_params.dummySendCycleDuration;
+		while (!g_b_exit) {
+			if (TicksTime::now() >= nextCycleStartTime) {
```
keep TT:now() in temp variable instead of calling the function twice

avnerbh reviewed (commented):
nice feature, please see my comments

Commit:
Fix AvnerB comments:

1. SwitchOnCycleDuration ,SwitchOff : Remove unnecessary variables names from some functions.
2. SwitchOnDummySend :
      Use pMsgRequest's buffer instead of local buffer.
      Keep TT:now() in temp variable instead of calling the function twice.
      Update sendto comment.
3. update DUMMY_SEND_FLAG comment.

Signed-off-by: Liran Oz <lirano@mellanox.com>

liranoz12 commented:
@rosenbaumalex @avnerbh  Please review
